*  Copyright 2025 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description : Add support for Recursive DNS server option.
* Version : v001
* Mandatory/Optional : Mandatory
* Impact if not applied : thread 1.4 test fails.
*/

Upstream-Status: Backport

commit df12cc8159a7a0fff4044fb0686201137666d9af
Author: Yang Song <yangsongcn@google.com>
Date:   Thu Mar 6 02:08:32 2025 +0000

    [nd6] Add support for Recursive DNS Server (RDNSS) option

diff --git a/src/core/net/nd6.cpp b/src/core/net/nd6.cpp
index 3e55126fd..28e0a85ab 100644
--- a/src/core/net/nd6.cpp
+++ b/src/core/net/nd6.cpp
@@ -170,6 +170,17 @@ uint8_t RouteInfoOption::OptionLengthForPrefix(uint8_t aPrefixLength)
     return length;
 }
 
+//----------------------------------------------------------------------------------------------------------------------
+// RecursiveDNSServer
+
+void RecursiveDNSServer::Init(void)
+{
+    Clear();
+    SetType(kTypeRaFlagsExtension);
+
+    OT_UNUSED_VARIABLE(mReserved);
+}
+
 //----------------------------------------------------------------------------------------------------------------------
 // RaFlagsExtOption
 
diff --git a/src/core/net/nd6.hpp b/src/core/net/nd6.hpp
index c16421ce4..7aeb4188f 100644
--- a/src/core/net/nd6.hpp
+++ b/src/core/net/nd6.hpp
@@ -78,11 +78,12 @@ class Option
 public:
     enum Type : uint8_t
     {
-        kSourceLinkLayerAddr  = 1,  ///< Source Link Layer Address Option.
-        kTargetLinkLayerAddr  = 2,  ///< Target Link Layer Address Option.
-        kTypePrefixInfo       = 3,  ///< Prefix Information Option.
-        kTypeRouteInfo        = 24, ///< Route Information Option.
-        kTypeRaFlagsExtension = 26, ///< RA Flags Extension Option.
+        kSourceLinkLayerAddr    = 1,  ///< Source Link Layer Address Option.
+        kTargetLinkLayerAddr    = 2,  ///< Target Link Layer Address Option.
+        kTypePrefixInfo         = 3,  ///< Prefix Information Option.
+        kTypeRouteInfo          = 24, ///< Route Information Option.
+        kTypeRecursiveDNSServer = 25, ///< Recursive DNS Server Option.
+        kTypeRaFlagsExtension   = 26, ///< RA Flags Extension Option.
     };
 
     static constexpr uint16_t kLengthUnit = 8; ///< The unit of length in octets.
@@ -445,6 +446,79 @@ private:
 
 static_assert(sizeof(RouteInfoOption) == 8, "invalid RouteInfoOption structure");
 
+/**
+ * Represents an Recursive DNS Server Option.
+ *
+ * See RFC-5175 [https://tools.ietf.org/html/rfc5175]
+ */
+OT_TOOL_PACKED_BEGIN
+class RecursiveDNSServer : public Option, private Clearable<RecursiveDNSServer>
+{
+    friend class Clearable<RecursiveDNSServer>;
+
+public:
+    static constexpr Type kType = kTypeRecursiveDNSServer; ///< RA Recursive DNS Server Option type.
+
+    /**
+     * See section 5,1 of RFC 8106 for definition of this option. [https://tools.ietf.org/html/rfc8106#section-5.1]
+     */
+    void Init(void);
+
+    /**
+     * Tells whether this option is valid.
+     *
+     * @returns  A boolean indicates whether this option is valid.
+     */
+    bool IsValid(void) const { return GetSize() >= sizeof(*this); }
+
+    /**
+     * Gets the lifetime of the DNS server in seconds.
+     *
+     * @returns The lifetime in seconds.
+     */
+    uint32_t GetLifetime(void) const { return BigEndian::HostSwap32(mLifetime); }
+
+    /**
+     * Gets the number of DNS server addresses in the option.
+     *
+     * @returns The number of DNS server addresses.
+     */
+    uint8_t GetNumAddresses(void) const { return (GetLength() > 1) ? (GetLength() - 1) / 2 : 0; }
+
+    /**
+     * Gets a pointer to the start of the DNS server addresses.
+     *
+     * @returns A pointer to the start of the DNS server addresses.
+     */
+    const Address *GetAddresses(void) const
+    {
+        return reinterpret_cast<const Address *>(reinterpret_cast<const uint8_t *>(this) + sizeof(*this));
+    }
+
+    RecursiveDNSServer(void) = delete;
+
+private:
+    // RA Recursive DNS Server Option
+    //
+    //   0                   1                   2                   3
+    //   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    //  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //  |     Type      |     Length    |           Reserved            |
+    //  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //  |                           Lifetime                            |
+    //  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //  |                                                               |
+    //  :            Addresses of IPv6 Recursive DNS Servers            :
+    //  |                                                               |
+    //  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                               .
+
+    uint16_t mReserved; // The reserved field.
+    uint32_t mLifetime; // The lifetime in seconds.
+    // Followed by Address array (variable length)
+} OT_TOOL_PACKED_END;
+
+static_assert(sizeof(RecursiveDNSServer) == 8, "invalid RecursiveDNSServer structure");
+
 /**
  * Represents an RA Flags Extension Option.
  *
